(Making Contributor Dashboard UI Responsive) Milestone 1.2: Implement the mobile UI for list item components

Lawful2002 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of N/A.
2. This PR does the following: Make list-item component responsive on all the tabs according to these mocks: [mocks](https://www.figma.com/file/gpS0ZJhT1KWrjwWrY9Ow3T/Contributor-Dashboard---With-Question-Submission-Workflow?node-id=0%3A1)

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

### Before
![Screenshot from 2022-06-24 23-02-10](https://user-images.githubusercontent.com/73544247/175612773-7c1b3d4e-1c9e-4ca6-9d24-103c31263e71.png)
![Screenshot from 2022-06-24 23-02-02](https://user-images.githubusercontent.com/73544247/175612778-797a9019-161a-48f2-9ca4-93e80bf162c7.png)


### After
#### Desktop (breakpoint > 700px)
![image](https://user-images.githubusercontent.com/73544247/177024178-53bd77ee-27bb-4104-969c-6f45f0bc485c.png)


![Screenshot from 2022-06-24 22-12-21](https://user-images.githubusercontent.com/73544247/175607480-f45b799e-a3e9-47f6-8021-959b3912f088.png)
![Screenshot from 2022-06-24 22-12-29](https://user-images.githubusercontent.com/73544247/175607486-826bc858-4bfc-4a70-8a24-234225079114.png)

#### Mobile
##### Breakpoint < 500px

![image](https://user-images.githubusercontent.com/73544247/177024197-8e6a5c9c-0839-4e1c-89ec-8d3bf625590d.png)

![image](https://user-images.githubusercontent.com/73544247/175976862-6d9eb738-b984-48f0-ac96-b113d3af6c0a.png)

![image](https://user-images.githubusercontent.com/73544247/175976977-16f74464-1c58-49b6-b1c5-95a484cf0200.png)


##### Breakpoint 500px - 700px
![image](https://user-images.githubusercontent.com/73544247/175977242-9194519a-0a28-47c5-92ee-125cb46d317c.png)
![image](https://user-images.githubusercontent.com/73544247/175977273-af364a40-508e-4b95-8f5b-2ce861c3f574.png)

![image](https://user-images.githubusercontent.com/73544247/177024214-0f136c3d-0695-4f64-8225-203988e8a216.png)


## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

vojtechjelinek committed:
fixed UI of list item components

Lawful2002 committed:
made UI for list-items responsive

Lawful2002 committed:
Merge remote-tracking branch 'upstream/develop' into milestone1.2

Lawful2002 committed:
fixed issues

Lawful2002 committed:
fixed issues

Lawful2002 committed:
fixed issues with files

Lawful2002 commented:
@vojtechjelinek PTAL.

Lawful2002 committed:
fixed class name

Lawful2002 committed:
Merge branch 'oppia:develop' into milestone1.2

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.html:
```diff
@@ -1,28 +1,61 @@
       </span>
+      <span title="{{ opportunity?.subheading }}"
+            class="protractor-test-opportunity-list-item-subheading oppia-subheading-small"
+            [hidden]="opportunityDataIsLoading">  {{ opportunity?.subheading | wrapTextWithEllipsis : 20 }}
```
```suggestion
            [hidden]="opportunityDataIsLoading">  
        {{ opportunity?.subheading | wrapTextWithEllipsis : 20 }}
```

vojtechjelinek replied:
ditto above

Lawful2002 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.html:
```diff
@@ -1,28 +1,61 @@
+        <div class="progress-bar bg-light"
+             role="progressbar"
+             [ngStyle]="untranslatedProgressStyle"
+             ngbTooltip="{{opportunity.totalCount - (opportunity.translationsCount + opportunity.inReviewCount)}}/{{opportunity.totalCount}} cards available for translation"
```
Can we move this count to the TS file? It seems a bit complicated for HTML.

vojtechjelinek replied:
ditto below

Lawful2002 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.html:
```diff
@@ -44,14 +77,128 @@
+              [hidden]="opportunityDataIsLoading || correspondingOpportunityDeleted"
+              [disabled]="opportunityButtonDisabled"
+              (click)="clickActionButton.emit(opportunity.id)">
+        {{ opportunity?.actionButtonTitle }}
```
Why `?`?

vojtechjelinek replied:
ditto elsewhere

Lawful2002 replied:
The `?` is needed to make sure that the 'opportunity' is not null/undefined before attempting to read the 'actionButtonTitle' property.

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.html:
```diff
@@ -44,14 +77,128 @@
+</div>
+
+<!-- Mobile -->
+<div *ngIf="showOpportunityButton"
```
So the mobile vs desktop is only decide using CSS rules?

Lawful2002 replied:
Yes, only CSS (media queries) are used to differentiate between Desktop and Mobile views.

vojtechjelinek replied:
Hmm, how does that affects load speed? Wouldn't it be better to hide and show these using `ngIf`?

Lawful2002 replied:
Fixed.

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass. Can you please include screenshots that contain different status of the suggestions? (other than pending review)

Lawful2002 committed:
fixed merge conflicts

Lawful2002 committed:
addressed changes

Lawful2002 commented:
@vojtechjelinek PTAL.

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.html:
```diff
@@ -62,7 +213,114 @@
+      <lazy-loading [hidden]="!opportunityDataIsLoading"></lazy-loading>
+    </div>
+    <div class="oppia-opportunities-heading oppia-animate-fadein">
+      <span class="protractor-test-opportunity-list-item-heading" title="{{ opportunity?.heading }}"
+            [hidden]="opportunityDataIsLoading">
```
```suggestion
      <span class="protractor-test-opportunity-list-item-heading" 
            title="{{ opportunity?.heading }}"
            [hidden]="opportunityDataIsLoading">
```

vojtechjelinek reviewed (changes requested):
Thanks! Only one coding issue. I'm thinking about the status ('accepted', 'review pending', etc.) part of each line. I think it would look a bit better if it was always the same size regardless of the text. Do you think that would be possible and better?

Lawful2002 committed:
Update core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.html

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

Lawful2002 commented:
> Thanks! Only one coding issue. I'm thinking about the status ('accepted', 'review pending', etc.) part of each line. I think it would look a bit better if it was always the same size regardless of the text. Do you think that would be possible and better?

@vojtechjelinek Setting a fixed width for these labels is causing issues with margin and alignment on some widths because the height of this label grows/shrinks according to screen size and setting a fixed width will not be reliable. Perhaps we could remove the background color of these labels altogether, something like this: 
![image](https://user-images.githubusercontent.com/73544247/177023691-d6af3943-2b95-4195-9558-39b85d544ab2.png)

Thoughts?

Lawful2002 committed:
addressed comments

Lawful2002 committed:
merged origin

Lawful2002 committed:
Merge remote-tracking branch 'origin/milestone1.2' into milestone1.2

Lawful2002 commented:
> Thanks! Only one coding issue. I'm thinking about the status ('accepted', 'review pending', etc.) part of each line. I think it would look a bit better if it was always the same size regardless of the text. Do you think that would be possible and better?

Fixed (I have updated the screenshots). PTAL.

Lawful2002 committed:
Merge branch 'oppia:develop' into milestone1.2

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.ts:
```diff
@@ -20,7 +20,10 @@ import { Component, Input, Output, EventEmitter } from '@angular/core';
 import { downgradeComponent } from '@angular/upgrade/static';
 
 import constants from 'assets/constants';
+import { floor } from 'mathjs';
```
Can you use native JS function?

Lawful2002 replied:
Fixed.

vojtechjelinek reviewed (changes requested):
Thanks! One nit.

Lawful2002 committed:
using native js functions

Lawful2002 committed:
Merge remote-tracking branch 'origin/milestone1.2' into milestone1.2

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.ts:
```diff
@@ -49,6 +56,8 @@ export class OpportunitiesListItemComponent {
   @Input() labelRequired: boolean = false;
   @Input() progressBarRequired: boolean = false;
 
+  @Input() showOpportunityButton: boolean = true;
```
Remove the empty line above.

Lawful2002 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.ts:
```diff
@@ -67,6 +79,12 @@ export class OpportunitiesListItemComponent {
   opportunityButtonDisabled: boolean = false;
 
   ngOnInit(): void {
+    this.onMobile = this.windowDimensionsService.getWidth() <= 700;
```
Please move the `700` into a constant and also add a comment explaining that it needs to be synchronized with the CSS styles (do the same in the CSS styles).

Lawful2002 replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Two other comments that I missed before.

Lawful2002 committed:
addressed changes

Lawful2002 committed:
addressed changes

Lawful2002 commented:
@vojtechjelinek @sagangwee @kevintab95 @gp201 @nithusha21 PTAL.

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.ts:
```diff
@@ -39,15 +41,20 @@ export interface ExplorationOpportunity {
 export class OpportunitiesListItemComponent {
+  constructor(
+    private windowDimensionsService: WindowDimensionsService
+  ) {
+  }
```
```suggestion
  ) {}
```

vojtechjelinek reviewed (approved):
Thanks! Just one nit.

Lawful2002 committed:
Update core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.ts

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

Lawful2002 commented:
@sagangwee @kevintab95 @gp201 @nithusha21 PTAL.

sagangwee commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.html:
```diff
@@ -44,14 +80,130 @@
+</div>
+
+<!-- Mobile -->
+<div *ngIf="showOpportunityButton && onMobile"
```
A lot of the child HTML components seem to be the same and so it will be difficult to make changes in the future—you'd have to check each "version", compare the diff and make changes in each version. Is it possible to extract the common parts into subcomponents?

Lawful2002 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.html:
```diff
@@ -44,14 +80,130 @@
+      <span title="{{ opportunity?.subheading }}"
+            class="protractor-test-opportunity-list-item-subheading oppia-subheading-small"
+            [hidden]="opportunityDataIsLoading">
+        {{ opportunity?.subheading | wrapTextWithEllipsis : 20 }}
```
The `wrapTextWithEllipsis` might need to be dynamically adjusted based on the breakpoint, since it looks like the text is truncated too early when breakpoint < 500px.

Lawful2002 replied:
As this component is used on multiple pages, setting a value according to a single page might cause issue elsewhere. I have increased the cutoff to 25 (instead of 20) so more of the text is visible without it wrapping over to the next line.

sagangwee commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.spec.ts:
```diff
@@ -20,13 +20,26 @@ import { ComponentFixture, fakeAsync, TestBed } from '@angular/core/testing';
+  }
+
+  getWidth(): number {
+    return 530;
```
Where does this number come from? Would be good to document in a comment.

Lawful2002 replied:
This can be any arbitrary number below 700 to simulate a mobile viewport. I have changed it to 390 (width for iPhone 12) to be a bit more specific.

sagangwee commented on core/templates/pages/contributor-dashboard-page/opportunities-list-item/opportunities-list-item.component.ts:
```diff
@@ -80,7 +98,8 @@ export class OpportunitiesListItemComponent {
       if (this.opportunity.progressPercentage) {
         this.progressPercentage = (
-          this.opportunity.progressPercentage + '%');
+          String((Math.floor(this.opportunity.progressPercentage))) + '%'
```
Could we use a template literal for this to avoid the String cast?

Lawful2002 replied:
Done

sagangwee commented on assets/constants.ts:
```diff
@@ -5899,6 +5899,11 @@ export default {
+  // The breakpoint for mobile view for contributor dashboard in px.
+  // This value must be the same as the one specified in
+  // opportunities-list-item.component.html.
+  "CONTRIBUTOR_DASHBOARD_MOBILE_BREAKPOINT": 700,
```
Will this only be relevant only for opportunities-list-item.component.html? If so, then let's update the constant name to reflect that, since currently it makes it sound like the breakpoint value is relevant for the entire contributor dashboard, which I don't think it is (e.g. I remember there being many different breakpoint values in the last PR).

Lawful2002 replied:
Fixed.

sagangwee reviewed (changes requested):
@Lawful2002 Thanks, took a pass.

Lawful2002 committed:
[skip ci] testing changes

kevintab95 reviewed (approved):
Same comment as Sagang in the codeowner file. LGTM otherwise. Thanks!

Lawful2002 committed:
addressed comments

Lawful2002 committed:
fixed merge conflicts

Lawful2002 commented:
@sagangwee PTAL.

gp201 reviewed (approved):
LGTM

Lawful2002 commented:
@sagangwee PTAL.

Lawful2002 committed:
Merge branch 'oppia:develop' into milestone1.2

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributions-and-review/contributions-and-review.component.html:
```diff
@@ -278,7 +283,7 @@
   }
 
-  @media only screen and (min-width: 601px) and (max-width: 1500px) {
+  @media only screen and (min-width: 701px) and (max-width: 1500px) {
```
Hm typically we shouldn't need both min and max width as long as the media queries are ordered correctly. Since we're developing desktop first, we should order by descending order of max width to be consistent.

Lawful2002 replied:
Fixed

sagangwee reviewed (approved):
@Lawful2002 Thanks, LGTM. Just one comment.

Lawful2002 committed:
fixed ordering of media queries

Lawful2002 committed:
Merge remote-tracking branch 'origin/milestone1.2' into milestone1.2

Lawful2002 committed:
fixed e2e test

Lawful2002 committed:
fixes

Lawful2002 committed:
Merge branch 'oppia:develop' into milestone1.2

Lawful2002 commented:
@SAEb-ai PTAL.

SAEb-ai commented on core/tests/protractor_desktop/contributorDashboard.js:
```diff
@@ -148,7 +148,7 @@ describe('Contributor dashboard page', function() {
     await contributorDashboardPage.waitForOpportunitiesToLoad();
     await contributorDashboardPage.expectOpportunityWithPropertiesToExist(
-      SKILL_DESCRIPTIONS[0], TOPIC_NAMES[0], null, '(0.00%)');
+      SKILL_DESCRIPTIONS[0], TOPIC_NAMES[0], null, '(0%)');
```
Why this change? Ditto elsewhere.

Lawful2002 replied:
This change is needed because now in the front-end we show percentages without decimals.

Lawful2002 commented:
@SAEb-ai PTAL.

Lawful2002 commented:
@SAEb-ai PTAL.

Lawful2002 committed:
Merge branch 'oppia:develop' into milestone1.2

Lawful2002 commented:
@SAEb-ai Could you take a look?

Lawful2002 committed:
Merge branch 'oppia:develop' into milestone1.2
